src/nuopc/cap.F90

```diff
@@ -0,0 +1,3131 @@
+module nexus_cap
+
+  !-----------------------------------------------------------------------------
+  ! NEXUS NUOPC Component.
+  !-----------------------------------------------------------------------------
+
+  use ESMF
+  use NUOPC
+  use NUOPC_Model, modelSS => SetServices
+
+  use HCO_STATE_MOD, only: Hco_State
+  use HCO_TYPES_MOD, only: ConfigObj
+  use HCO_Error_Mod, only: rk_hco => hp, &
+    HCO_LogFile_Open, &
+    HCO_Error, HCO_MSG, HCO_Leave, HCO_Enter, &
+    HCO_SUCCESS, HCO_MISSVAL
+  use HCOX_STATE_MOD, only: Ext_State
+
+  implicit none
+
+  ! TODO: cap object with pointers that can be retrieved with ESMF_GridCompGetInternalState?
+
+  ! Default values for HEMCO input files: contain definitions of
+  ! species, grid, and time settings, etc.
+  character(len=255) :: GridFile = 'HEMCO_sa_Grid'
+  character(len=255) :: SpecFile = 'HEMCO_sa_Spec'
+  character(len=255) :: TimeFile = 'HEMCO_sa_Time'
+  character(len=255) :: DiagFile = 'NEXUS_Diag.nc'
+  character(len=255) :: ExptFile = 'NEXUS_Expt.nc'
+
+  !> HEMCO config object
+  type(ConfigObj), pointer :: HcoConfig => NULL()
+
+  !> HEMCO state object
+  type(Hco_State), pointer :: HcoState => NULL()
+
+  !> HEMCO extensions state
+  type(Ext_State), pointer :: HcoExtState => NULL()
+
+  type(ESMF_Grid)  :: HCO_Grid
+  type(ESMF_Grid)  :: NXS_Grid
+  type(ESMF_State) :: NXS_Diag_State
+  !! "importState"
+  !! An ESMF state of diagnostics on the HEMCO grid.
+  type(ESMF_State) :: NXS_Expt_State
+  !! "exportState"
+  !! Regridded to the desired output grid.
+  type(ESMF_RouteHandle) :: NXS_RouteHandle
+
+  logical :: do_Regrid = .false.
+  !! True if grid file path passed to `init` is not empty string.
+  logical :: do_Debug  = .false.
+  !! True if `debugLevel` passed to `init` is greater than zero.
+  logical :: do_NEXUS  = .false.
+  !! True if either `do_Regrid` or `do_Debug` is true.
+  logical :: alwaysWriteRestartFile = .false.
+  !! Even in NEXUS mode (`do_NEXUS`)
+
+  ! Start and end time of simulation
+  integer :: T_YY(2), T_MM(2), T_DD(2)
+  integer :: T_H(2), T_M(2), T_S(2)
+
+  ! Grid
+  real(rk_hco), allocatable, target :: XMID   (:,:,:)
+  real(rk_hco), allocatable, target :: YMID   (:,:,:)
+  real(rk_hco), allocatable, target :: XEDGE  (:,:,:)
+  real(rk_hco), allocatable, target :: YEDGE  (:,:,:)
+  real(rk_hco), allocatable, target :: YSIN   (:,:,:)
+  real(rk_hco), allocatable, target :: AREA_M2(:,:,:)
+  real(rk_hco), allocatable, target :: PBL_M  (:,:)
+
+  !> MAXIT is the maximum number of run calls allowed
+  integer, parameter :: MAXIT = 100000
+
+  integer, parameter :: rootPet = 0
+
+  type(ESMF_StaggerLoc), parameter :: staggerList(2) = &
+    (/ ESMF_STAGGERLOC_CENTER, ESMF_STAGGERLOC_CORNER /)
+
+  private
+
+  public SetServices, &
+    init, finalize, &
```

Comment:
Maybe these should be `nxs_init` and `nxs_finalize`

Reply:
I like that.

Reply:
Unresolving until I change them

Reply:
Changed in 4a20c97